src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,684 @@
+    // concurrently initializing the same instance (delayed provider selection)
+    // in a way which would corrupt the internal state.
+    private final Object lock = new Object();
+
+
+    // Instantiates a {@code KDF} object. This constructor is called when a
+    // provider is supplied to {@code getInstance}.
+    //
+    // @param delegate the delegate
+    // @param algorithm the algorithm
+    // @param kdfParameters the parameters
+    private KDF(Delegate delegate, String algorithm, KDFParameters kdfParameters) {
+        this.pairOfSpiAndProv = delegate;
+        this.algorithm = algorithm;
+        // note that the parameters are being passed to the impl in getInstance
+        this.kdfParameters = kdfParameters;
+        serviceIterator = null;
+    }
+
+    // Instantiates a {@code KDF} object. This constructor is called when a
+    // provider is not supplied to {@code getInstance}.
+    //
+    // @param firstPairOfSpiAndProv the delegate
+    // @param t the service iterator
+    // @param algorithm the algorithm
+    // @param kdfParameters the algorithm parameters
+    private KDF(Delegate firstPairOfSpiAndProv, Iterator<Service> t, String algorithm,
+                KDFParameters kdfParameters) {
+        this.firstPairOfSpiAndProv = firstPairOfSpiAndProv;
+        serviceIterator = t;
+        this.algorithm = algorithm;
+        this.kdfParameters = kdfParameters;
+    }
+
+    /**
+     * Returns the algorithm name of this {@code KDF} object.
+     *
+     * @return the algorithm name of this {@code KDF} object
+     */
+    public String getAlgorithm() {
+        return this.algorithm;
+    }
+
+    /**
+     * Returns the name of the provider.
+     *
+     * @see <a href="#DelayedProviderSelection">Delayed Provider Selection</a>
+     *
+     * @return the name of the provider
+     */
+    public String getProviderName() {
+        useFirstSpi();
+        return pairOfSpiAndProv.provider().getName();
+    }
+
+    /**
+     * Returns the {@code KDFParameters} used with this {@code KDF} object.
+     * <p>
+     * The returned parameters may be the same that were used to initialize
+     * this {@code KDF} object, or may contain additional default or
+     * random parameter values used by the underlying KDF algorithm.
+     * If the required parameters were not supplied and can be generated by
+     * the {@code KDF} object, the generated parameters are returned;
+     * otherwise {@code null} is returned.
+     *
+     * @see <a href="#DelayedProviderSelection">Delayed Provider Selection</a>
+     *
+     * @return the parameters used with this {@code KDF} object, or
+     * {@code null}
+     */
+    public KDFParameters getParameters() {
+        useFirstSpi();
+        return pairOfSpiAndProv.spi().engineGetParameters();
+    }
+
+    /**
+     * Returns a {@code KDF} object that implements the specified algorithm.
+     *
+     * @see <a href="#DelayedProviderSelection">Delayed Provider Selection</a>
```

Comment:
Put `@see` at the end. Put `@implNote` before `@param`. This is the order they are rendered in HTML javadoc.